File: php/BedrockSlimeFinder.php
<?
class MersenneTwister {
    const N = 624;
    const M = 397;
    const MATRIX_A = 0x9908B0DF;
    const UPPER_MASK = 0x80000000;
    const LOWER_MASK = 0x7FFFFFFF;
    
    private $mt;
    private $mti;
    
    public function __construct($seed) {
        $this->mt = array_fill(0, self::N, 0);
        $this->mti = self::N + 1;
        $this->initSeed($seed);
    }
    
    private function initSeed($s) {
        $this->mt[0] = $s & 0xFFFFFFFF;
        for ($i = 1; $i < self::N; $i++) {
            $t = $this->mt[$i - 1] ^ (($this->mt[$i - 1] >> 30) & 0xFFFFFFFF);
            $this->mt[$i] = ((1812433253 * ((($t & 0xFFFF0000) >> 16) & 0xFFFF)) << 16) + 
                            (1812433253 * ($t & 0xFFFF)) + $i;
            $this->mt[$i] = $this->mt[$i] & 0xFFFFFFFF;
        }
        $this->mti = self::N;
    }
    
    public function randomInt() {
        $mag01 = [0x0, self::MATRIX_A];
        
        if ($this->mti >= self::N) {
            if ($this->mti == self::N + 1) {
                $this->initSeed(5489);
            }
            
            $kk = 0;
            while ($kk < self::N - self::M) {
                $y = ($this->mt[$kk] & self::UPPER_MASK) | ($this->mt[$kk + 1] & self::LOWER_MASK);
                $y_shifted = ($y >> 1) & 0x7FFFFFFF;
                $this->mt[$kk] = $this->mt[$kk + self::M] ^ $y_shifted ^ $mag01[$y & 0x1];
                $kk++;
            }
            while ($kk < self::N - 1) {
                $y = ($this->mt[$kk] & self::UPPER_MASK) | ($this->mt[$kk + 1] & self::LOWER_MASK);
                $y_shifted = ($y >> 1) & 0x7FFFFFFF;
                $this->mt[$kk] = $this->mt[$kk + (self::M - self::N)] ^ $y_shifted ^ $mag01[$y & 0x1];
                $kk++;
            }
            $y = ($this->mt[self::N - 1] & self::UPPER_MASK) | ($this->mt[0] & self::LOWER_MASK);
            $y_shifted = ($y >> 1) & 0x7FFFFFFF;
            $this->mt[self::N - 1] = $this->mt[self::M - 1] ^ $y_shifted ^ $mag01[$y & 0x1];
            $this->mti = 0;
        }
        
        $y = $this->mt[$this->mti++];
        $y = $y ^ (($y >> 11) & 0xFFFFFFFF);
        $y = $y ^ ((($y << 7) & 0xFFFFFFFF) & 0x9D2C5680);
        $y = $y ^ ((($y << 15) & 0xFFFFFFFF) & 0xEFC60000);
        $y = $y ^ (($y >> 18) & 0xFFFFFFFF);
        
        return $y & 0xFFFFFFFF;
    }
}

class BedrockSlimeFinder {
    private static function umul32_lo($a, $b) {
        $a = $a & 0xFFFFFFFF;
        $b = $b & 0xFFFFFFFF;
        
        $a_low = $a & 0xFFFF;
        $b_low = $b & 0xFFFF;
        $result = $a_low * $b_low;
        $carry = ($result >> 16) & 0xFFFF;
        
        $carry += ($a >> 16) * $b_low;
        $carry &= 0xFFFF;
        
        $carry += $a_low * ($b >> 16);
        
        return ((($carry & 0xFFFF) << 16) | ($result & 0xFFFF)) & 0xFFFFFFFF;
    }
    
    public static function getSlimeRandomSeed($x, $z) {
        $seed = self::umul32_lo($x & 0xFFFFFFFF, 522133279) ^ ($z & 0xFFFFFFFF);
        $rng = new MersenneTwister($seed);
        return $rng->randomInt();
    }
    
    public static function isSlimeChunk($x, $z) {
        $randomVal = self::getSlimeRandomSeed($x, $z);
        return ($randomVal % 10) === 0;
    }
}

File: c++/BedrockSlimeFinder.cpp
#include "BedrockSlimeFinder.h"

uint64_t BedrockSlimeFinder::umul32_lo(uint64_t a, uint64_t b) {
    uint64_t a_low = a & 0xFFFFLL;
    uint64_t b_low = b & 0xFFFFLL;
    uint64_t result = a_low * b_low;
    uint64_t carry = result >> 16;

    carry += (a >> 16) * b_low;
    carry &= 0xFFFFLL;

    carry += a_low * (b >> 16);

    return ((carry & 0xFFFFLL) << 16 | (result & 0xFFFFLL)) & 0xFFFFFFFFLL;
}
void BedrockSlimeFinder::MersenneTwister::initSeed(uint32_t s) {
    mt[0] = s & 0xFFFFFFFFU;
    for (mti = 1; mti < N; mti++) {
        uint32_t t = mt[mti - 1] ^ (mt[mti - 1] >> 30);
        mt[mti] = (((1812433253U * ((t & 0xFFFF0000U) >> 16)) << 16) +
                   (1812433253U * (t & 0xFFFFU)) + mti) & 0xFFFFFFFFU;
    }
}

BedrockSlimeFinder::MersenneTwister::MersenneTwister(uint32_t seed) {
    mti = N + 1;
    initSeed(seed);
};

uint32_t BedrockSlimeFinder::MersenneTwister::randomInt() {
    uint32_t mag01[2] = {0x0U, MATRIX_A};

    if (mti >= N) {
        int kk;
        if (mti == N + 1) initSeed(5489U);

        for (kk = 0; kk < N - M; kk++) {
            uint32_t y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
            mt[kk] = mt[kk + M] ^ (y >> 1) ^ mag01[y & 0x1U];
        }
        for (; kk < N - 1; kk++) {
            uint32_t y = (mt[kk] & UPPER_MASK) | (mt[kk + 1] & LOWER_MASK);
            mt[kk] = mt[kk + (M - N)] ^ (y >> 1) ^ mag01[y & 0x1U];
        }
        uint32_t y = (mt[N - 1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
        mt[N - 1] = mt[M - 1] ^ (y >> 1) ^ mag01[y & 0x1U];
        mti = 0;
    }

    uint32_t y = mt[mti++];
    y ^= y >> 11;
    y ^= (y << 7) & 0x9D2C5680U;
    y ^= (y << 15) & 0xEFC60000U;
    y ^= y >> 18;

    return y & 0xFFFFFFFFU;
}


uint32_t BedrockSlimeFinder::getSlimeRandomSeed(int x, int z) {
    uint64_t seed = umul32_lo(static_cast<uint32_t>(x), 522133279U) ^ static_cast<uint32_t>(z);
    MersenneTwister rng(static_cast<uint32_t>(seed));
    return rng.randomInt();
}

bool BedrockSlimeFinder::isSlimeChunk(int x, int z) {
    uint32_t randomVal = getSlimeRandomSeed(x, z);
    return (randomVal % 10) == 0;
}